### PR TITLE
Fix annotation and upgrade django

### LINKED ISSUE #4: Unexpected default value with fallback annotations
Hi :-)

I'm currently trying to wrap my head around the default behavior of certain aggregations/annotations like Sum and noticed that with the GFK fallback query there is a slightly different outcome than I had anticipated.

If I, for instance, have a food (to stick the the models used in the test cases) that has zero ratings, according to Django's normal behavior this should result in a score of None.

But in in the updated tests (see below) the test for the new food "peach" fails with the value 0 for sqlite3 or 0L for postgres.

```
diff --git a/generic_aggregation/generic_aggregation_tests/tests.py b/generic_aggregation/generic_aggregation_tests/tests.py
index 89aee97..2346a67 100644
--- a/generic_aggregation/generic_aggregation_tests/tests.py
+++ b/generic_aggregation/generic_aggregation_tests/tests.py
@@ -14,6 +14,7 @@ class SimpleTest(TestCase):
     def setUp(self):
         self.apple = Food.objects.create(name='apple')
         self.orange = Food.objects.create(name='orange')
+        self.peach = Food.objects.create(name='peach')
         
         dt = datetime.datetime(2010, 1, 1)
         
@@ -36,7 +37,7 @@ class SimpleTest(TestCase):
         return _generic_filter(*args, **kwargs)
         
     def test_annotation(self):
-        annotated_qs = self.generic_annotate(Food.objects.all(), Rating.objects.all(), models.Count('ratings__rating'))
+        annotated_qs = self.generic_annotate(Food.objects.all(), Rating.objects.all(), models.Count('ratings__rating')).order_by('-score')
         self.assertEqual(annotated_qs.count(), 2)
         
         food_a, food_b = annotated_qs
@@ -121,15 +122,18 @@ class SimpleTest(TestCase):
         o1 = CharFieldGFK.objects.create(name='o1', content_object=self.orange)
         
         annotated_qs = self.generic_annotate(Food.objects.all(), CharFieldGFK, models.Count('char_gfk__name'))
-        self.assertEqual(annotated_qs.count(), 2)
+        self.assertEqual(annotated_qs.count(), 3)
         
-        food_a, food_b = annotated_qs.order_by('-score')
+        food_a, food_b, food_c = annotated_qs.order_by('-score')
         
         self.assertEqual(food_b.score, 1)
         self.assertEqual(food_b.name, 'orange')
         
         self.assertEqual(food_a.score, 2)
         self.assertEqual(food_a.name, 'apple')
+
+        self.assertEqual(food_c.score, None)
+        self.assertEqual(food_c.name, 'peach')
     
         aggregated = self.generic_aggregate(Food.objects.all(), CharFieldGFK, models.Count('char_gfk__name'))
         self.assertEqual(aggregated, 3)
```

Is this an expected outcome and did I misunderstand something? :-)

### LINKED ISSUE #6: Add Django 1.8, 1.9 compatability
and remove less then 1.7

### PR DESCRIPTION
**Pull request by jdavisp3**
@coleifer @auvipy @zerok Fixes #4 Fixes #6

This PR is intended to fix #4 (unexpected fallback annotations) and #6 (newer Django compatibility).

The annotations appeared to have two problems:

  * For ORM-based annotation, there were missing results for rows without any matching generic relations. Standard Django annotation and aggregation includes those results so I think the fallback version is more correct here. I don't think ORM-based annotation can be easily made equivalent to the fallback so this PR uses the fallback-based annotation exclusively.
  * For fallback annotation, the default value when there were no matching rows was ending up `0` always because of the `COALESCE` call, even for aggregate functions that would normally return `null`. Removing `COALESCE` results in the same value as the standard ORM-based annotation for the aggregate function being used.
  * Fixing this bug uncovered a test bug which assumed `null` values are always sorted to the front on ascending sorts. However, postgres uses the opposite convention as sqlite. This is fixed  with a test setting which the test can use to determine the expected order from test queries.

The Django upgrades required the following changes.
  * Django 1.8 changed the API on aggregates for obtaining lookup information. This is fixed by using the newer API for 1.8 onwards.
  * Django 1.8 syncs apps without migrations before other apps. This caused a problem since the test app has a foreign key to the `ContentType` table. This is fixed by adding migrations to the test app. There is a new settings file for use in updating migrations in the future if the test app models need to change.
  * Django 1.9 removed the older import path for `GenericForeignKey`. This new import path is supported in 1.7 and 1.8 so this is fixed by just using the new path.
  * Django 1.9 fails because importing `ContentType` fails before the apps are all loaded, and the test app is underneath the main package so loading triggers that import. This is fixed by moving the test app out of the main package.

This PR also drops support for pre-1.7 versions, updates the docs, adds a HISTORY file, and bumps the version to `0.4.0`.

How this was tested
------------------

* A clean `tox` run.

### CODE REVIEW

File: generic_aggregation_tests/tests.py [hunk 1]
```diff
@@ -167,7 +205,7 @@ def test_filter_cast(self):
             self.assertEqual(obj.content_object.name, 'apple')
 
 class FallbackTestCase(SimpleTest):
-    def generic_aggregate(self, *args, **kwargs):
+    def generic_annotate(self, *args, **kwargs):
```
Comment (jdavisp3) [inline generic_aggregation_tests/tests.py]: This is why the fallback version appeared to be equivalent to the normal version in tests -- it wasn't being tested.

Comment (coleifer) [inline generic_aggregation_tests/tests.py]: d'ungh

File: generic_aggregation_tests/tests.py [hunk 2]
```diff
@@ -111,6 +132,19 @@ def test_subset_annotation(self):
         self.assertEqual(food_b.score, 7)
         self.assertEqual(food_b.name, 'orange')
 
+        self.assertEqual(food_c.score, None)
+        self.assertEqual(food_c.name, 'peach')
+
+        persimmon = Food.objects.create(name='persimmon')
```
Comment (coleifer) [inline generic_aggregation_tests/tests.py]: Persimmons?!

**Comment by coleifer** [comment]
Fantastic work, thanks for this excellent PR. Feel free to contribute to all my projects in the future.

**Comment by coleifer** [comment]
I'll push 0.4 to pypi

File: generic_aggregation_tests/tests.py (hunk 2 above)
Comment (jdavisp3) [inline generic_aggregation_tests/tests.py]: My second favorite fruit after figs :smile:

**Comment by jdavisp3** [comment]
Thanks @coleifer !

**Comment by jdavisp3** [comment]
And great to work with you again :smiley:

**Comment by coleifer** [comment]
No, thank *you*, thanks for keeping this project alive.

**Comment by zerok** [comment]
@jdavisp3 Thank you so much, that's awesome! :smiley:

**Pull request merged**